packages/graph/migrations/postgres/migration/1656417312397_initial.ts

```diff
@@ -341,168 +330,353 @@ export const up = (pgm: MigrationBuilder): void => {
-    )`,
-    },
+      -- insert the data of the entity
+      INSERT INTO entity_editions (
+        updated_by_id,
+        archived,
+        entity_type_version_id,
+        properties,
+        left_to_right_order,
+        right_to_left_order
+      ) VALUES (
+        _updated_by_id,
+        _archived,
+        _entity_type_version_id,
+        _properties,
+        _left_to_right_order,
+        _right_to_left_order
+      ) RETURNING entity_editions.entity_edition_id INTO _entity_edition_id;
+
+      RETURN QUERY
+      INSERT INTO entity_versions (
+        owned_by_id,
+        entity_uuid,
+        entity_edition_id,
+        decision_time,
+        system_time
+      ) VALUES (
+        _owned_by_id,
+        _entity_uuid,
+        _entity_edition_id,
+        tstzrange(_decision_time, 'infinity'),
+        tstzrange(now(), 'infinity')
+      ) RETURNING entity_versions.entity_edition_id, entity_versions.decision_time, entity_versions.system_time;
+    END
+    `,
   );
 
-  pgm.addConstraint(
-    "entity_histories",
-    "entities_histories_relation_order_constraint",
+  pgm.createFunction(
+    "update_entity",
+    [
+      {
+        name: "_owned_by_id",
+        type: "UUID",
+      },
+      {
+        name: "_entity_uuid",
+        type: "UUID",
+      },
+      {
+        name: "_decision_time",
+        type: "TIMESTAMP WITH TIME ZONE",
+      },
+      {
+        name: "_updated_by_id",
+        type: "UUID",
+      },
+      {
+        name: "_archived",
+        type: "BOOLEAN",
+      },
+      {
+        name: "_entity_type_version_id",
+        type: "UUID",
+      },
+      {
+        name: "_properties",
+        type: "JSONB",
+      },
+      {
+        name: "_left_to_right_order",
+        type: "INTEGER",
+      },
+      {
+        name: "_right_to_left_order",
+        type: "INTEGER",
+      },
+    ],
     {
-      // Because of the "entities_histories_relation_constraint", we can check any one of the required link columns
-      check: `(left_entity_uuid IS NOT NULL)
-            OR (left_to_right_order IS NULL AND right_to_left_order IS NULL)`,
+      returns:
+        "TABLE (entity_edition_id BIGINT, decision_time tstzrange, system_time tstzrange)",
+      language: "plpgsql",
+      replace: true,
     },
+    `
+    DECLARE
+      _new_entity_edition_id BIGINT;
+    BEGIN
+      -- If the decision time is not specified, use the current time.
+      IF _decision_time IS NULL THEN _decision_time := now(); END IF;
+  
+      -- ensure, that the row, we are going to update, is locked
+      -- this is required as \`UPDATE\` will not reevaluate the \`WHERE\` clause
+      PERFORM FROM entity_versions
+      WHERE entity_versions.owned_by_id = _owned_by_id
+        AND entity_versions.entity_uuid = _entity_uuid
+        AND entity_versions.decision_time @> _decision_time
+        AND entity_versions.system_time @> now()
+      FOR UPDATE;
```

Comment:
Discussed offline:
Because of the way the Read Commited isolation level works (default one we're using), we can expect the locks to read the updated row and re-running the condition, but not to rescan for any newly inserted rows to re-lock. This would mean that in the concurrent case of two sessions updating on top of the same row, we would see the last session failing its update (returning 0 insertions) if it tries to change for a decision time before the first session. This happens because the row would not pass the where clause the second time around (decision time would be different)

To get around this, we decided we should be able to get rid of this `for update` clause and rely on the update locking at the bottom, and then in application code retry the entity update if the number of *inserted rows* is 0. We would do that by rolling back the transaction in that case.